### ISSUE TITLE
Returned instance type is wrong when using Generics multiple times 

### ISSUE THREAD

**Issue by oboenikui** [Kotlin/kotlinx.coroutines#1723]
Kotlin: 1.3.61
Coroutines: 1.3.3

When I use coroutines with generics, returned instance type is wrong in some cases.
In the following cases, String is returned at the place where Unit should be returned.

```kotlin
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.test.runBlockingTest
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test

class CoroutinesTest {

    @Test
    fun `TestUseCase should return Unit`() = runBlockingTest {
        val uc = TestUseCase()
        val result: Either<Error, Unit> = uc()
        // assertion failed because result.right is String
        assertThat(result.right).isInstanceOf(Unit::class.java)
    }
}

class TestUseCase : BuggyUseCase<Unit>() {

    override suspend fun run() {
        doSomething()
        // WORKAROUND: store to some variable
        // val ignored = doSomething()
    }

    private suspend fun doSomething(): String {
        return coroutineScope {
            // this bug occurs when the coroutine scope suspended at least once
            delay(1)
            "hello"
        }
    }
}

abstract class BuggyUseCase<T> {
    protected abstract suspend fun run(): T

    // I defined Either class for proof, but any classes using generics occur same problem.
    suspend operator fun invoke(): Either<Error, T> {
        // run() should return Unit on TestUseCase, but actual is "hello"
        return Either.right(run())
    }
}

sealed class Either<L, R> {
    abstract val left: L?
    abstract val right: R?

    class Left<L, R>(override val left: L) : Either<L, R>() {
        override val right: R? = null
    }

    class Right<L, R>(override val right: R) : Either<L, R>() {
        override val left: L? = null
    }

    companion object {
        fun <L, R> left(value: L): Left<L, R> = Left(value)
        fun <L, R> right(value: R): Right<L, R> = Right(value)
    }
}
```

Thank you.

**Comment by qwwdfsad** [comment]
This is a compiler bug that will be fixed in 1.4.
Please see #1678

**Comment by oboenikui** [comment]
I'm sorry, I couldn't find the issue. Thank you!
